lib/auth/auth_with_roles.go

```diff
@@ -5627,7 +5702,49 @@ func (a *ServerWithRoles) DeleteAllSAMLIdPServiceProviders(ctx context.Context)
 		return trace.Wrap(err)
 	}
 
-	return a.authServer.DeleteAllSAMLIdPServiceProviders(ctx)
+	// Get all of the service names and entity IDs so that we can emit the delete events for this later.
+	nameAndEntityID := map[string]string{}
+	var nextKey string
+	for {
+		var sps []types.SAMLIdPServiceProvider
+		var err error
+		sps, nextKey, err = a.authServer.ListSAMLIdPServiceProviders(ctx, 0, nextKey)
+		if err != nil {
+			return trace.Wrap(err)
```

Comment:
If we return here or after the `DeleteAllSAMLIdPServiceProviders` call fails, we won't emit the audit event, right?

Reply:
That's right. I'll correct this.